Introduce an auto deploy feature

tiwanari opened the pull request:
resolves #20.

This PR introduces a feature of auto deploy.

### Procedure
1. Merge changes into `master` (e.g., `git checkout master; git merge develop`)
1. Edit the version of the project in `build.gradle` ([this line](https://github.com/PileProject/drivecommand/blob/37a4b6f7f83d76a6de077aedb7052be701df719f/build.gradle#L63))
1. Run a task (`./gradlew uploadArchives`)
1. Make a commit (e.g., `git add .; git commit -m "Release v2.2.0"`)
1. Make a tag (e.g., `git tag v2.2.0`)
1. Push changes (`git push --follow-tags`)

Then, `gh-pages` branch will be updated automatically :tada:

### References
- [promises-guide/deploy-gh-pages.sh at master · w3ctag/promises-guide]( https://goo.gl/6816am )
- (Japanese) [Travis CI経由でgh-pagesにmiddlemanで構築した静的サイトをデプロイ。 - Qiita]( https://goo.gl/GxNSIq )

tiwanari committed:
Introduce an auto deploy feature

mandaiy commented:
I don't know this magical things but seems ok.
Did you test this?

tiwanari commented:
Yes, I tested the procedure with my forked repository and it worked. But, strictly speaking, I haven't tested this PR because the `secure` was newly created for this repository. I hope it works properly 🙏 

#### TL; DR
I copied the script and the changes on `.travis.yml` for this PR, and made a new `personal access token` on my https://github.com/settings/tokens page by checking only `Access public repositories`.

After getting the token, I ran the following command for this repository:
```
travis encrypt -r PileProject/drivecommand "GH_TOKEN=TOKEN" --add
```
and this added some lines into `.travis.yml` automatically. The result is this PR.

tiwanari commented:
By the way, I'm not sure whether `master` branch should have [release directory](https://github.com/PileProject/drivecommand/tree/master/com/pileproject/drivecommand). And, I found it now diversed from `gh-pages` branch's [release directory](https://github.com/PileProject/drivecommand/tree/gh-pages/com/pileproject/drivecommand). What do you think? :)

tiwanari commented:
Ah, I remembered that the uploadArchives task refers to the existing releases by checking [maven metadata file](https://github.com/PileProject/drivecommand/blob/master/com/pileproject/drivecommand/maven-metadata.xml). So, at least we should keep the file in master and keep (restore) the consistence.

tiwanari commented:
Hey @makotoshimazu @myusak 

Please give me comments or approvals:) I want to merge this and make `v2.2.0` before merging #23 and making `v2.3.0` tag for that.

mandaiy commented on scripts/deploy-gh-pages.sh:
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
```
I think we should remove these comments to avoid needless copyright violation and rewrite our own words.

tiwanari replied:
👍

mandaiy commented on scripts/deploy-gh-pages.sh:
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
+set -e # Exit with nonzero exit code if anything fails
+
+mkdir tmp
```
Maybe it's safer way to use `mktemp` than `mkdir`, though `mktemp` is not in posix but I'm sure that the running OS is Ubuntu.

tiwanari replied:
Thanks for the information:)

mandaiy commented on scripts/deploy-gh-pages.sh:
```diff
@@ -0,0 +1,53 @@
+TARGET_BRANCH="gh-pages"
+
+# Pull requests and commits to other branches shouldn't try to deploy, just build to verify
+if [ "$TRAVIS_PULL_REQUEST" != "false" -o -z "$TRAVIS_TAG" ]; then
```
I'd split this `if` statement and add messages why we are skipping deploy, like `echo 'Skipping deploy because $TRAVIS_TAG is missing'`
I don't know where you define these environmental variables so if my comment is pointless please ignore.

tiwanari replied:
The list of variables is [here](https://docs.travis-ci.com/user/environment-variables/#Default-Environment-Variables) and you can see "`TRAVIS_TAG: If the current build is for a git tag, this variable is set to the tag’s name.`" So, I found that I can remove this compound because I specify the trigger [here](https://github.com/PileProject/drivecommand/pull/22/files#diff-354f30a63fb0907d4ad57269548329e3R24) and that requires a tag (and thus, it cannot be a PR).

mandaiy commented on scripts/deploy-gh-pages.sh:
```diff
@@ -0,0 +1,53 @@
+# Clone the existing gh-pages for this repo into out/
+# Create a new empty branch if gh-pages doesn't exist yet (should only happen on first deply)
+git clone $REPO src
+cd src; git checkout $SOURCE_BRANCH
```
I don't feel the necessity of `;`, why not remove this and write two separate commands?

tiwanari replied:
Okay:)

mandaiy commented on scripts/deploy-gh-pages.sh:
```diff
@@ -0,0 +1,53 @@
+# Build
+./gradlew uploadArchives
+
+cd ..
```
I think we should avoid `cd`'ing with relative paths.
Keep the root path of this temporary directory and use it for `cd`

tiwanari replied:
👍

mandaiy commented on scripts/deploy-gh-pages.sh:
```diff
@@ -0,0 +1,53 @@
+git config --global user.email "travis@travis-ci.org"
+git config --global user.name "Travis"
+
+# If there are no changes (e.g. this is a README update) then just bail.
```
Maybe we can check this earlier, just below the `cd tgt`

tiwanari replied:
I'm not sure but stuff like `git config foo` should be done before `git diff`(?) but anyway I will refactor this file.

mandaiy commented on scripts/deploy-gh-pages.sh:
```diff
@@ -0,0 +1,53 @@
+# The delta will show diffs between new and old versions.
+git add .
+git commit -m "Release ${TRAVIS_TAG}"
+git push --quiet $REPO_WITH_TOKEN $TARGET_BRANCH
```
Better to echo a message to tell we've done the deploy successfully?

tiwanari replied:
👍

mandaiy commented:
I had a detailed look on your changes and now I have a few questions and comments.

Questions:
- How can we specify `$TRAVIS_TAG`?
Is it already defined in Travis?

Comments:
- I investigated `build.gradle` and I found a way to specify `uploadArchives.repositories.mavenDeployers.pom.version` via command line arguments.
If you have time please change your script (L26) and then we don't have to change `build.gradle` per our release.
With my diff, the artifacts can be generated by `./gradlew deploy -Ptag=$TAG`
```diff
diff --git a/build.gradle b/build.gradle
index a2bc4b0..c8b36ab 100644
--- a/build.gradle
+++ b/build.gradle
@@ -44,10 +44,6 @@ dependencies {
 
 }
 
-task wraper(type: Wrapper) {
-    gradleVersion = '2.5'
-}
-
 // for our license
 license {
     header rootProject.file('LICENSE')
@@ -55,26 +51,37 @@ license {
 }
 
 // Maven repository identity setting
-uploadArchives {
-    repositories {
-        mavenDeployer {
-            repository url: "file:${projectDir}"
-
-            pom.version = '2.1.0'
-            pom.groupId = 'com.pileproject'
-            pom.artifactId = 'drivecommand'
+uploadArchives.repositories.mavenDeployer {
+    repository url: "file:${projectDir}"
+    pom {
+        version = 'unspecified'
+        groupId = 'com.pileproject'
+        artifactId = 'drivecommand'
 
-            pom.project {
-                inceptionYear '2011'
-                packaging 'jar'
-                licenses {
-                    license {
-                        name 'The Apache Software License, Version 2.0'
-                        url 'http://www.apache.org/licenses/LICENSE-2.0.txt'
-                        distribution 'repo'
-                    }
+        project {
+            inceptionYear '2011'
+            packaging 'jar'
+            licenses {
+                license {
+                    name 'The Apache Software License, Version 2.0'
+                    url 'http://www.apache.org/licenses/LICENSE-2.0.txt'
+                    distribution 'repo'
                 }
             }
         }
     }
 }
+
+task deploy {
+    description = "Deploys current sourceset into .jar file via 'uploadArchives' task."
+
+    doLast {
+        if (!project.hasProperty('tag')) {
+            throw new GradleException('Tag not specified. Run with -Ptag=$TAG')
+        }
+
+        uploadArchives.repositories.mavenDeployer.pom.version = project.tag
+
+        uploadArchives.execute()
+    }
+}
```

- Now we can publish GitHub pages via `/docs` directory in `master` branch
https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/
So your script could be simplified because we don't have to deal with chaining branches.

mandaiy reviewed (changes requested):
Please check my comments in your script.
As I wrote in conversation, if you have time please consider enhancing your script.

tiwanari reviewed (commented):
Thank you for your review, @myusak !
They are really helpful 👍 

I will fix them and refactor scripts. Please give me a little time:)

tiwanari committed:
Apply myusak's comments

tiwanari committed:
Fix deploy script

tiwanari commented:
Hi @myusak 

I fixed scripts and tested them:)

### Update
Thanks to your valuable comments, the procedure becomes much simpler 🎉 

1. Merge changes into `master` (e.g., `git checkout master; git merge develop`)
~1. Edit the version of the project in `build.gradle` ([this line](https://github.com/PileProject/drivecommand/blob/37a4b6f7f83d76a6de077aedb7052be701df719f/build.gradle#L63))~
~1. Run a task (`./gradlew uploadArchives`)~
~1. Make a commit (e.g., `git add .; git commit -m "Release v2.2.0"`)~
1. Make a tag (e.g., `git tag v2.2.0`)
1. Push changes (`git push --follow-tags`)

### Note
I specified the format of a tag for deploy to `^v[0-9].*$` and added a feature removing the prefix `v` in the task you suggested to follow the current naming rule of release folders.

### Verification
[This tag commit](https://travis-ci.org/tiwanari/drivecommand#L299) successfully deployed the release to `/docs` and [this commit](https://github.com/tiwanari/drive/commit/f17be04a61c8713e51d6c0da56e83e367da8fedb) properly changed the source of the library in [CI](https://travis-ci.org/tiwanari/drive#L2340) (and it has errors because the `v.2.2.0` changed the signature of a function).

tiwanari commented on .travis.yml:
```diff
@@ -1,18 +1,25 @@
+  on:
+    branch: master
+    tags: true
+    condition: "$TRAVIS_TAG =~ ^v[0-9].*$"
```
Check the format here.

tiwanari replied:
Check the format here.

tiwanari commented on build.gradle:
```diff
@@ -44,37 +44,49 @@ dependencies {
+            throw new GradleException('Tag not specified. Run with -Ptag=$TAG')
+        }
+
+        if (!(project.tag =~ /^v\d.*$/)) {
```
This is the same check of `.travis.yaml`.

tiwanari commented on build.gradle:
```diff
@@ -44,37 +44,49 @@ dependencies {
+        }
+
+        // remove the prefix 'v'
+        uploadArchives.repositories.mavenDeployer.pom.version = project.tag.substring(1)
```
remove `v`

tiwanari reviewed (commented):
Added a few comments. 

Please review this again, @myusak

mandaiy reviewed (approved):
Looks good. Good job 👍

tiwanari committed:
Add a dependency for deploy task

tiwanari commented:
Thanks, @myusak !! :)

Let me merge it 🎉
